app/controllers/history_controller.rb

```diff
@@ -0,0 +1,58 @@
+class HistoryController < LinksControllerBase
+  before_action :authorize, only: %i[update]
+
+  # 2. set @link instance var, since a lot of action filters use it
+  before_action do
+    set_link(params[:link_id])
+  end
+  before_action :require_link
+  before_action only: %i[show update] do
+    set_past_link(params[:id])
+  end
+  before_action :show, only: %i[show]
+
+  # 3. protect link-specific buisness rules
+  before_action :past_link_matches_link, only: %i[show update]
+  before_action :prevent_public_expired, if: -> { link_is_expired_and_not_owned? }
+  before_action :skip_unauthorized_requests, only: %i[update], if: -> { !past_link_belongs_to_user? }
+  before_action :protect_friends_only_links, only: %i[index show update]
+
+  after_action :track_visit, only: %i[index show]
+
+  def index
+    @past_links = PastLink.all.order(id: :desc).where(link: @link).take(50)
+  end
+
+  def show
```

Comment:
One of the consequences of showing the modal like this is that users can go directly to the history entry's page by clicking "open in new tab" or going to the URL directly. From my googling, it seemed like this is something that's considered normal and desirable in Rails.

The resulting page works completely and I've tested it separately to confirm that everything works exactly as intended on the individual page. I made sure to organize the Stimulus controllers in such a way that they worked here too.